### PR TITLE
Add the option to define a separate ResourceGroup for the VNET in Azure

### LINKED ISSUE #836: Add the option to define a seperate ResourceGroup for the VNET in the Azure Cloud Provider
### Is your proposal related to a problem?

If the VNET in Azure is located in a different Ressource Group than the machines, the deployment of worker cluster is failing. The reason for this it, it is lookking for the network in the given Resource Group and it is not possible to add an optional RessourceGroup for the VNET.

### Describe the solution you'd like

The perfect solution would be to have a optional parameter vnetResourceGroup that if it is filled it uses the ResourceGroup that is given. If none is given just use the ResourceGroup wich is defined before. 

### Additional context

I created a gist with the changes i guess need to be made. Also i'm not shure if it is working and if the default option is working by falling back to the normal RessourceGroup if the VNetRessourceGroup is not given.
[You can find the gist here](https://gist.github.com/mfahlandt/70284ade3f5ad45b0803ac889d0bbeeb)

### PR DESCRIPTION
**Pull request by medmedchiheb**
**What this PR does / why we need it**:
VNET Fixes #836: [Azure] Add VnetResourceGroup option for VNET

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #836

**Special notes for your reviewer**:

**Optional Release Note**:

```release-note
Separate ResourceGroup for the VNET in Azure
```

### CODE REVIEW

**Comment by kron4eg** [comment]
Hi @medmedchiheb and thank you for the PR!

/test all

**Comment by kron4eg** [comment]
@medmedchiheb looks like this PR misses the `go mod tidy` invocation.

File: pkg/cloudprovider/provider/azure/types/cloudconfig.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ type CloudConfig struct {
 	RouteTableName             string  `json:"routeTableName"`
 	SecurityGroupName          string  `json:"securityGroupName" yaml:"securityGroupName"`
 	PrimaryAvailabilitySetName string  `json:"primaryAvailabilitySetName"`
-	VnetResourceGroup          *string `json:"vnetResourceGroup,omitempty"`
+	VnetResourceGroup          string  `json:"vnetResourceGroup"`
```
Comment (kron4eg) [inline pkg/cloudprovider/provider/azure/types/cloudconfig.go:38]:
```
INFO [runner] linters took 16.147422301s with stages: goanalysis_metalinter: 14.441121508s, unused: 1.692719018s
pkg/cloudprovider/provider/azure/types/cloudconfig.go:31: File is not `gofmt`-ed with `-s` (gofmt)
        ResourceGroup              string  `json:"resourceGroup"`
        Location                   string  `json:"location"`
        VNetName                   string  `json:"vnetName"`
        SubnetName                 string  `json:"subnetName"`
        RouteTableName             string  `json:"routeTableName"`
        SecurityGroupName          string  `json:"securityGroupName" yaml:"securityGroupName"`
        PrimaryAvailabilitySetName string  `json:"primaryAvailabilitySetName"`
        VnetResourceGroup          string  `json:"vnetResourceGroup"`
        UseInstanceMetadata        bool    `json:"useInstanceMetadata"`
```

Comment (medmedchiheb) [reply pkg/cloudprovider/provider/azure/types/cloudconfig.go]: sorry

Comment (kron4eg) [reply pkg/cloudprovider/provider/azure/types/cloudconfig.go]: No worries, that's why we have CI and lint automated :D

**Comment by kron4eg** [comment]
@medmedchiheb seems like the fallback didn't worked.
from the test log: https://public-prow.loodse.com/view/gs/prow-dev-public-data/pr-logs/pull/kubermatic_machine-controller/852/pull-machine-controller-e2e-azure/1333361953313132545

```
        helper.go:204: verify failed due to error=failed to create machine migrateuid-azure-1333361953313132545: failed to generate main network interface: failed to fetch subnet: network.SubnetsClient#Get: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="" Message="No HTTP resource was found that matches the request URI 'https://management.azure.com/subscriptions/528474a9-8cc6-46c8-9b74-1dbc31c4bb7f/resourceGroups/providers/Microsoft.Network/virtualNetworks/machine-controller-e2e/subnets/machine-controller-e2e?api-version=2018-06-01'."
```

**Comment by kron4eg** [comment]
@medmedchiheb I've fixed small typo in the title :)

File: pkg/cloudprovider/provider/azure/provider.go [hunk 2]
```diff
@@ -827,6 +834,10 @@ func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
 		return errors.New("resourceGroup is missing")
 	}
 
+	if c.VNetResourceGroup == "" {
```
Comment (moadqassem) [inline pkg/cloudprovider/provider/azure/provider.go:837]: This is not needed as you are already setting the `VNetResourceGroup` in here https://github.com/medmedchiheb/machine-controller/blob/azure-vnet-resource-group/pkg/cloudprovider/provider/azure/provider.go#L224

File: pkg/cloudprovider/provider/azure/types/cloudconfig.go [hunk 3]
```diff
@@ -28,15 +28,15 @@ type CloudConfig struct {
...
 	AADClientSecret string `json:"aadClientSecret"`
 
-	ResourceGroup              string  `json:"resourceGroup"`
-	Location                   string  `json:"location"`
-	VNetName                   string  `json:"vnetName"`
-	SubnetName                 string  `json:"subnetName"`
-	RouteTableName             string  `json:"routeTableName"`
-	SecurityGroupName          string  `json:"securityGroupName" yaml:"securityGroupName"`
-	PrimaryAvailabilitySetName string  `json:"primaryAvailabilitySetName"`
-	VnetResourceGroup          *string `json:"vnetResourceGroup,omitempty"`
```
Comment (moadqassem) [inline pkg/cloudprovider/provider/azure/types/cloudconfig.go:38]: Why ?

**Comment by moadqassem** [comment]
@medmedchiheb Please add the new introduced field in the e2e tests too:
https://github.com/medmedchiheb/machine-controller/blob/azure-vnet-resource-group/test/e2e/provisioning/testdata/machinedeployment-azure.yaml#L26

https://github.com/medmedchiheb/machine-controller/blob/azure-vnet-resource-group/test/e2e/provisioning/testdata/machinedeployment-azure-redhat-satellite.yaml#L26

**Comment by moadqassem** [comment]
@medmedchiheb Would you like to finish addressing the code review? or shall we just take over as this is ticket is kinda blocker. Thanks

**Comment by medmedchiheb** [comment]
@moadqassem  I'm sorry I was on vacation, and sorry because I didn't receive the notification in my work email.
So, I notice that you have added it yourself in master, anything else needed to do here?

**Comment by moadqassem** [comment]
@medmedchiheb This ticket was actually a blocker for a high-prio ticket, that's why we took over it. At this point the issue has been address in the other PR so I would suggest that we close this one :)

**Comment by toschneck** [comment]
for the ref. #874

**Pull request closed without merging**